### ISSUE TITLE
index produces panic: utf8_mg_pos_cache_update

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10455]
Migrated from [rt.perl.org#75898](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=75898) (status was 'resolved')

Searchable as RT75898$

**Comment by p5pRT** [comment]
### From @cpansprout

This may be related to 69422\, but it has nothing to do with overloading&#8203;:

perl5\.13\.1 \-le' utf8&#8203;::upgrade\($\_ = "   "\); print index $\_\, " "\, 76'
panic&#8203;: utf8\_mg\_pos\_cache\_update cache 76 real 3 for     at \-e line 1\.

This only happens with a debugging build\.

This is the cause of \<https://rt.cpan.org/Ticket/Display.html?id=58371>.


\-\-\-\-Program\-\-\-\-
\#\!perl \-l
BEGIN\{$\{^UTF8CACHE\}=\-1\}
utf8&#8203;::upgrade\($\_ = "   "\); print index $\_\, " "\, 4

\-\-\-\-Output of \.\.\./pQTLY7N/perl\-5\.8\.0@&#8203;27574/bin/perl\-\-\-\-
\-1

\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of \.\.\./p8Y5JFT/perl\-5\.8\.0@&#8203;27575/bin/perl\-\-\-\-
panic&#8203;: utf8\_mg\_pos\_cache\_update cache 4 real 3 for     at test2 line 3\.

\-\-\-\-EOF \($?='65280'\)\-\-\-\-
Need a perl between 27574 and 27575
No patch available between 27574 and 27575

http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/0905937




---
Flags&#8203;:
&nbsp;     category=core
&nbsp;     severity=low


---
Site configuration information for perl 5\.13\.1&#8203;:

Configured by sprout at Tue Jun 15 06&#8203;:25&#8203;:06 PDT 2010\.

Summary of my perl5 \(revision 5 version 13 subversion 1 patch  
v5\.13\.1\-196\-gee84a79\) configuration&#8203;:
&nbsp;   Snapshot of&#8203;: ee84a7925f473937ece8b2db88bffc4400a8cabe
&nbsp;   Platform&#8203;:
&nbsp;     osname=darwin\, osvers=10\.0\.0\, archname=darwin\-2level
&nbsp;     uname='darwin pint\.local 10\.0\.0 darwin kernel version 10\.0\.0&#8203;: fri  
jul 31 22&#8203;:47&#8203;:34 pdt 2009; root&#8203;:xnu\-1456\.1\.25~1release\_i386 i386 '
&nbsp;     config\_args='\-de \-Dusedevel \-DDEBUGGING'
&nbsp;     hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;     useithreads=undef\, usemultiplicity=undef
&nbsp;     useperlio=define\, d\_sfio=undef\, uselargefiles=define\,  
usesocks=undef
&nbsp;     use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;     usemymalloc=n\, bincompat5005=undef
&nbsp;   Compiler&#8203;:
&nbsp;     cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-no\-cpp\-precomp \- 
DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/ 
include'\,
&nbsp;     optimize='\-O3 \-g'\,
&nbsp;     cppflags='\-no\-cpp\-precomp \-fno\-common \-DPERL\_DARWIN \-no\-cpp\- 
precomp \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/ 
usr/local/include'
&nbsp;     ccversion=''\, gccversion='4\.2\.1 \(Apple Inc\. build 5646\)'\,  
gccosandvers=''
&nbsp;     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\,  
Off\_t='off\_t'\, lseeksize=8
&nbsp;     alignbytes=8\, prototype=define
&nbsp;   Linker and Libraries&#8203;:
&nbsp;     ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-fstack\- 
protector \-L/usr/local/lib'
&nbsp;     libpth=/usr/local/lib /usr/lib
&nbsp;     libs=\-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;     perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;     libc=/usr/lib/libc\.dylib\, so=dylib\, useshrplib=false\,  
libperl=libperl\.a
&nbsp;     gnulibc\_version=''
&nbsp;   Dynamic Linking&#8203;:
&nbsp;     dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;     cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/ 
usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:




---
@&#8203;INC for perl 5\.13\.1&#8203;:
&nbsp;     /usr/local/lib/perl5/site\_perl/5\.13\.1/darwin\-2level
&nbsp;     /usr/local/lib/perl5/site\_perl/5\.13\.1
&nbsp;     /usr/local/lib/perl5/5\.13\.1/darwin\-2level
&nbsp;     /usr/local/lib/perl5/5\.13\.1
&nbsp;     /usr/local/lib/perl5/site\_perl
&nbsp;     \.



---
Environment for perl 5\.13\.1&#8203;:
&nbsp;     DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;     HOME=/Users/sprout
&nbsp;     LANG=en\_US\.UTF\-8
&nbsp;     LANGUAGE \(unset\)
&nbsp;     LD\_LIBRARY\_PATH \(unset\)
&nbsp;     LOGDIR \(unset\)
&nbsp;     PATH=/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/ 
usr/local/bin
&nbsp;     PERL\_BADLANG \(unset\)
&nbsp;     SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Jun 20\, 2010 at 12&#8203;:59&#8203;:50PM \-0700\, Father Chrysostomos wrote&#8203;:

> perl5\.13\.1 \-le' utf8&#8203;::upgrade\($\_ = "   "\); print index $\_\, " "\, 76'
> panic&#8203;: utf8\_mg\_pos\_cache\_update cache 76 real 3 for     at \-e line 1\.
> 
> This only happens with a debugging build\.

Or if you enable the cache consistency checks with \-C256 on the command line\.

\(Thanks for the report\)

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, Jun 21\, 2010 at 10&#8203;:11&#8203;:20AM \+0100\, Nicholas Clark wrote&#8203;:
> On Sun\, Jun 20\, 2010 at 12&#8203;:59&#8203;:50PM \-0700\, Father Chrysostomos wrote&#8203;:
> 
> > perl5\.13\.1 \-le' utf8&#8203;::upgrade\($\_ = "   "\); print index $\_\, " "\, 76'
> > panic&#8203;: utf8\_mg\_pos\_cache\_update cache 76 real 3 for     at \-e line 1\.
> > 
> > This only happens with a debugging build\.
> 
> Or if you enable the cache consistency checks with \-C256 on the command line\.

It's because S\_sv\_pos\_u2b\_forwards\(\) has no way to report that it hit the
end of the string before it reached the Unicode character position desired\.
\(And in turn\, S\_sv\_pos\_u2b\_midway\(\) needs this\)

I believe\, from inspection\, that the bug was present in the original
implementation of the caching code\, back in 5\.8\.1

http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/perl\-5\.8\.1&#8203;:/sv\.c\#l5732

However\, that implementation didn't have an optional cache consistency
check\, or the ability to disable caching\. \(See $\{^UTF8CACHE\} in perlvar\)

I'm not sure what the best way to fix this is\. I think we want to record
where the end of the string is\, if we fall off then end in this fashion\.

It's also not documented what the return value should be\. However\, this
isn't a huge problem as we should be replacing it with a
Perl\_sv\_pos\_b2u\_flags\(\)\, as part of fixing I32s that should be STRLENs\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, Jun 21\, 2010 at 01&#8203;:50&#8203;:45PM \+0100\, Nicholas Clark wrote&#8203;:

> It's because S\_sv\_pos\_u2b\_forwards\(\) has no way to report that it hit the
> end of the string before it reached the Unicode character position desired\.
> \(And in turn\, S\_sv\_pos\_u2b\_midway\(\) needs this\)
> 
> I believe\, from inspection\, that the bug was present in the original
> implementation of the caching code\, back in 5\.8\.1
> 
> http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/perl\-5\.8\.1&#8203;:/sv\.c\#l5732
> 
> However\, that implementation didn't have an optional cache consistency
> check\, or the ability to disable caching\. \(See $\{^UTF8CACHE\} in perlvar\)
> 
> I'm not sure what the best way to fix this is\. I think we want to record
> where the end of the string is\, if we fall off then end in this fashion\.

Assuming no\-one else says "Stop\!" in the next 16 hours\, I'll \(likely\) have a
go at fixing it over the weekend \(whilst offline\.\)

Nicholas Clark

**Comment by p5pRT** [comment]
### From @gannett-ggreer

On Mon Jun 21 05&#8203;:51&#8203;:12 2010\, nicholas wrote&#8203;:
> On Mon\, Jun 21\, 2010 at 10&#8203;:11&#8203;:20AM \+0100\, Nicholas Clark wrote&#8203;:
> > On Sun\, Jun 20\, 2010 at 12&#8203;:59&#8203;:50PM \-0700\, Father Chrysostomos wrote&#8203;:
> >
> > > perl5\.13\.1 \-le' utf8&#8203;::upgrade\($\_ = "   "\); print index $\_\, " "\,
> 76'
> > > panic&#8203;: utf8\_mg\_pos\_cache\_update cache 76 real 3 for     at \-e line
> 1\.
> > >
> > > This only happens with a debugging build\.
> >
> > Or if you enable the cache consistency checks with \-C256 on the
> command line\.
> 
> It's because S\_sv\_pos\_u2b\_forwards\(\) has no way to report that it hit
> the
> end of the string before it reached the Unicode character position
> desired\.
> \(And in turn\, S\_sv\_pos\_u2b\_midway\(\) needs this\)
> 
> I believe\, from inspection\, that the bug was present in the original
> implementation of the caching code\, back in 5\.8\.1
> 
> http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/perl\-5\.8\.1&#8203;:/sv\.c\#l5732
> 
> However\, that implementation didn't have an optional cache consistency
> check\, or the ability to disable caching\. \(See $\{^UTF8CACHE\} in
> perlvar\)
> 
> I'm not sure what the best way to fix this is\. I think we want to
> record
> where the end of the string is\, if we fall off then end in this
> fashion\.
> 
> It's also not documented what the return value should be\. However\,
> this
> isn't a huge problem as we should be replacing it with a
> Perl\_sv\_pos\_b2u\_flags\(\)\, as part of fixing I32s that should be
> STRLENs\.
> 
> Nicholas Clark
> 

Two possible solutions I came up with&#8203;:

A\) Maintain cache for offsets that go beyond the end of the string&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/sv.c b/sv.c
index c38a318..3140ec1 100644
--- a/sv.c
+++ b/sv.c
@@ -6386,7 +6386,7 @@ S_utf8_mg_pos_cache_update(pTHX_ SV *const sv,

```
</details>
MAGIC **const mgp, const STRLEN b
&nbsp;     \}
&nbsp;     assert\(cache\);
 
\-    if \(PL\_utf8cache \< 0 && SvPOKp\(sv\)\) \{
\+    if \(PL\_utf8cache \< 0 && SvPOKp\(sv\) && byte \< blen\) \{ /\* RT\#75898&#8203;:
We could be trying to offset past the end of the string\. \*/
&nbsp;     /\* SvPOKp\(\) because it's possible that sv has string overloading\, and
&nbsp;        therefore is a reference\, hence SvPVX\(\) is actually a pointer\.
&nbsp;        This cures the \(very real\) symptoms of RT 69422\, but I'm not actually

B\) Do not cache offsets that go beyond the end of the string\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/sv.c b/sv.c
index c38a318..841f238 100644
--- a/sv.c
+++ b/sv.c
@@ -6238,7 +6238,7 @@ S_sv_pos_u2b_cached(pTHX_ SV *const sv, MAGIC

```
</details>
**const mgp, const U8 *const start
&nbsp;     boffset = real\_boffset;
&nbsp;     \}
 
\-    if \(PL\_utf8cache\)
\+    if \(PL\_utf8cache && start \+ boffset \!= send\) /\* RT\#75898&#8203;: We could
be trying to offset past the end of the string\. \*/
&nbsp;     utf8\_mg\_pos\_cache\_update\(sv\, mgp\, boffset\, uoffset\, send \- start\);
&nbsp;     return boffset;
 \}

Both versions pass all the existing tests\.  And here's a test&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/t/uni/cache.t b/t/uni/cache.t
index df12f33..eb5b223 100644
--- a/t/uni/cache.t
+++ b/t/uni/cache.t
@@ -4,7 +4,7 @@ BEGIN {
     require "test.pl";
 }
 
-plan tests => 1;
+plan tests => 2;
 
 my $count = 0;
 unshift @INC, sub {
@@ -20,3 +20,8 @@ $s =~ m/[\p{Hiragana}]/;
 $s =~ m/[\p{Hiragana}]/;
 
 is($count, 1, "Swatch hash caching kept us from reloading swatch hash.");
+
+
+# RT#75898
+is(eval { utf8::upgrade($_ = " "); index $_, " ", 72 }, -1, "UTF-8

```
</details>
cache handles offset beyond the end of the string");
\+

Output before&#8203;:

1\.\.2
ok 1 \- Swatch hash caching kept us from reloading swatch hash\.
not ok 2 \- UTF\-8 cache handles offset beyond the end of the string
\# Failed at t/uni/cache\.t line 26
\#      got undef
\# expected "\-1"

Output after&#8203;:

1\.\.2
ok 1 \- Swatch hash caching kept us from reloading swatch hash\.
ok 2 \- UTF\-8 cache handles offset beyond the end of the string

\(The above were intentionally not sent as git commits so as to not lull
people into a false sense of thinking I know what I'm doing\.\) :\)

\-\- 
George Greer

**Comment by p5pRT** [comment]
### From @rgs

Fixed by :


commit 3e2d3818e517e0037c1ab6a482f31d50271f9e27
Author&#8203;: Nicholas Clark \<nick@&#8203;ccl4\.org>
Date&#8203;:   Sun Jul 11 20&#8203;:11&#8203;:10 2010 \+0100

&nbsp;    Avoid UTF\-8 cache panics with offsets beyond the string\. Fixes RT \#75898\.
&nbsp;    
&nbsp;    Change S\_sv\_pos\_u2b\_forwards\(\) to take a point to the \(requested\) UTF\-8 offset\,
&nbsp;    and return the actual UTF\-8 offset for the byte position returned\. This ensures
&nbsp;    that the cache is consistent with reality\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
